apps/users.py

```diff
@@ -106,6 +111,71 @@ def logout():
     return redirect(url_for('base.main'))
 
 
+class SignupForm(Form):
+    email = EmailField('Email', [Email(), Required()])
+    name = StringField('Name', [Required()])
+    allow_promo = BooleanField('Send me occasional emails about future EMF events')
+    signup = SubmitField('Sign up')
+
+    def validate_email(form, field):
+        if User.does_user_exist(field.data):
+            field.was_duplicate = True
+
+            msg = Markup(
+                render_template_string('Account already exists. '
+                                       'Please <a href="{{ url }}">click here</a> to log in.',
+                                       url=url_for('users.login', email=field.data)))
+            raise ValidationError(msg)
+
+
+@users.route("/signup", methods=['GET', 'POST'])
+def signup():
+    if not request.args.get('code'):
+        abort(404)
+
+    uid = verify_signup_code(app.config['SECRET_KEY'], time.time(), request.args.get('code'))
+    if uid is None:
+        flash("Your signup link was invalid. Please note that they expire after 6 hours.")
+        return redirect(url_for('.account'))
```

Comment:
Shouldn't this redirect to a login page rather than the account page?

Reply:
it will, indirectly. I'll make that more explicit